src/bluesearch/database/topic_rule.py

```diff
@@ -0,0 +1,125 @@
+# Blue Brain Search is a text mining toolbox focused on scientific use cases.
+#
+# Copyright (C) 2020  Blue Brain Project, EPFL.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Implementation of the TopicRule data structure."""
+from __future__ import annotations
+
+import re
+from typing import Iterable, Optional
+
+import pydantic
+
+from bluesearch.database.article import ArticleSource
+from bluesearch.database.topic_info import TopicInfo
+
+
+class TopicRule(pydantic.BaseModel, arbitrary_types_allowed=True):
+    """The topic rule."""
+
+    # None always represent wildcards
+    level: Optional[str] = None  # "article" or "journal"
+    source: Optional[ArticleSource] = None  # "arxiv", ... , "pubmed"
+    pattern: Optional[re.Pattern] = None  # regex pattern to match
+
+    @pydantic.validator("level")
+    def check_level_value(cls, value):
+        """Check the level parameter value."""
+        if value is not None and value not in {"article", "journal"}:
+            raise ValueError(f"Unsupported level {value}")
+        return value
+
+    @pydantic.validator("source", pre=True)
+    def convert_source_to_article_source(cls, value) -> ArticleSource | None:
+        """Check the source parameter value."""
+        if value is not None:
+            try:
+                value = ArticleSource(value)
+            except ValueError:
+                raise ValueError(f"Unsupported source {value}") from None
+
+        return value
+
+    @pydantic.validator("pattern", pre=True)
+    def convert_pattern_to_re(cls, value) -> re.Pattern | None:
+        """Check the pattern parameter value."""
+        if value is not None:
+            try:
+                value = re.compile(value)
+            except re.error:
+                raise ValueError(f"Unsupported pattern {value}") from None
+
+        return value
```

Comment:
@Stannislav probably is the right person to answer.

IMO the alternative would be doing the validation inside of `__post_init__` (https://www.python.org/dev/peps/pep-0557/#post-init-processing)

Reply:
I can try to answer this: 
Initially, we started the implementation with a `__post_init__` solution. One of the issue we had was because of `mypy`. 
For example, `source` was defined with type `str | ArticleSource | None` and directly converted to `ArticleSource | None` in the `__post_init__` but `mypy` was not processing that and complaining about the types in the other methods of the class.

Reply:
Yes, so `pydantic` was a bit of an experiment to see how'd it turn out. We can totally go for an alternative solution. We already had a solution (with a problem, see below) using dataclasses, so the commit that introduced `pydantic` can simply be reverted.

Regarding dependencies, many packages already depend on `pydantic` anyway, e.g. `spacy`, so we'll have it installed regardless. But yes, it's a new dependency.

The actualy motivation wasn't actually about nice error messages, even though I agree it's one of the prominent features of pydantic. The trouble was with type validation. Like @jankrepl said we used dataclasses and did parameter validation/conversion in `__post_init__`. So we had something like this:

```python
@dataclass
class TopicRule:
    """The topic rule."""
    pattern: str | re.Pattern | None = None

    def __post_init__(self) -> None:
        """Validate inputs."""
        if self.pattern is not None:
            try:
                self.pattern = re.compile(self.pattern)
            except re.error:
                raise ValueError(f"Unsupported pattern {self.pattern}") from None

    def match(self, topic_info: TopicInfo) -> bool:
        """Determine whether a topic_info matches the rule."""
        ...

        if self.pattern is None:
            return True
        ...
        self.pattern.match(...)
```
In `__post_init__` we always compile the patter if it's a string, so it's always of type `re.Pattern`, but mypy isn't clever enough to see that and complains that in `TopicRule.match` it could still be a `str`.

Obviously there are solutions in dataclasses - maybe not accept `str` as type for pattern. Or calling `re.compile` in `TopicRule.match`.

So in the end the problem with typing was a trigger to try something new. Given the purpose of `pydantic` is exactly to read configurations from disk/strings and do type validation/coersion it was kind of natural to go for it.

Reply:
On the data type: To be honest I am not even sure that we have a strong advantage in making it a `dataclass`.
I think the following is easy to read and doesn't need to handle any strange corner case with `mypy` or similar:
```python
class TopicRule:
    def __init__(
        self,
        level: str | None,
        source: str | ArticleSource | None,
        pattern: str | re.pattern | None,
    ):
        if level is not None and level not in {"article", "journal"}:
            raise ValueError(f"Unsupported level {level}.")
        self.level = level
        self.source = ArticleSource(source)
        self.pattern = re.compile(pattern)

    def match(self, topic_info: TopicInfo) -> bool:
        ...
```
Notice that exceptions are automatically raised here should `re.compile()` or `ArticleSource()` fail, even if the message isn't as pretty as in `pydantic`.

---

About `pydantic` specifically: Nothing against it, but I am bit afraid that our code starts growing a bit inorganically. I remember similar discussions about having valid use cases for `click`, but IIRC we decided to stick with `ArgParse` for consistency with the rest of the code and to avoid adding a dependency.

Reply:
Yes, good point, normal classes exist! 🚀 😄 . Seems legit too, probably would need to check for `None`, but otherwise LGTM.